controllers/clustercache/cluster_cache.go

```diff
@@ -169,34 +169,74 @@ type ClusterCache interface {
 var ErrClusterNotConnected = errors.New("connection to the workload cluster is down")
 
-// Watcher is a scoped-down interface from Controller that only has the Watch func.
+// Watcher is an interface that can start a Watch.
 type Watcher interface {
-	// Watch watches the provided Source.
-	Watch(src source.Source) error
+	Name() string
+	Object() client.Object
+	Watch(cache cache.Cache) error
+}
+
+// SourceWatcher is a scoped-down interface from Controller that only has the Watch func.
+type SourceWatcher[request comparable] interface {
+	Watch(src source.TypedSource[request]) error
 }
 
 // WatchInput specifies the parameters used to establish a new watch for a workload cluster.
-// A source.Kind source (configured with Kind, EventHandler and Predicates) will be added to the Watcher.
-// To watch for events, the source.Kind will create an informer on the Cache that we have created and cached
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the Watcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+type WatchInput = TypedWatchInput[client.Object, ctrl.Request]
+
+// TypedWatchInput specifies the parameters used to establish a new watch for a workload cluster.
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the Watcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
 // for the given Cluster.
-type WatchInput struct {
+type TypedWatchInput[object client.Object, request comparable] struct {
 	// Name represents a unique Watch request for the specified Cluster.
 	// The name is used to track that a specific watch is only added once to a cache.
 	// After a connection (and thus also the cache) has been re-created, watches have to be added
 	// again by calling the Watch method again.
 	Name string
 
 	// Watcher is the watcher (controller) whose Reconcile() function will be called for events.
-	Watcher Watcher
+	Watcher SourceWatcher[request]
 
 	// Kind is the type of resource to watch.
-	Kind client.Object
+	Kind object
 
 	// EventHandler contains the event handlers to invoke for resource events.
-	EventHandler handler.EventHandler
+	EventHandler handler.TypedEventHandler[object, request]
 
 	// Predicates is used to filter resource events.
-	Predicates []predicate.Predicate
+	Predicates []predicate.TypedPredicate[object]
+}
+
+// NewTypedWatcher creates a Watcher on the workload cluster.
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the SourceWatcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+func NewTypedWatcher[object client.Object, request comparable](input TypedWatchInput[object, request]) Watcher {
```

Comment:
```suggestion
func NewWatcher[object client.Object, request comparable](input TypedWatchInput[object, request]) Watcher {
```
Just simpler. I would only use Typed in something like this if we have to do it to avoid naming collisions